man/man1/xcat.1
.TH XCAT 1 "16th April 2015" "PUPSP3 commands" "PUPSP3 commands"
.SH NAME
.br

xcat \ the dynamic pipeline insertion point marker 
.SH SYNOPSIS
.B xcat 
[
.B -nice
.I cardinal
][
.B -argf
.I argument_file
][
.B -growth_protection
][
.B -help
][
.B -usage
][
.B -version
][
.B -slots
][
.B -slotinfo
]
!< <stdin>!
[1> <stdout>]
[2> <log file>]
.br

.SH DESCRIPTION
.I xcat 
provides a mechanism for inserting processes (and sub-pipelines) into a running
process pipeline dynamically. Typically, a process wishing to insert itself
into a running pipeline would open a PSRP channel to
.I xcat(1)
and then replac
.I xcat(1)
in the pipeline via the PSRP
.B overlay
(builtin) command. This facility allows the user to dynamically insert monitoring
filters (which may require transient facilities like an X display) into a running
pipeline.
.br
 
.SH OPTIONS
 
.B -help
invokes the on-line context help system for the file system watcher. This
"man" page is displayed. The process then exits.
.br

.B -usage
displays shortform usage string.
.br

.B -growth_protection
.I <retrys>
causes the system to retry an
.I sbrk(3)
call up to
.I retrys
times. This option permits the program to wait until the core that it requires
is available rather than simply exiting if
.I sbrk(3)
fails.
.br

.B -version
displays program version number.
.br

.B -slots
displays short form
.I pups(3)
library dependencies.
.br

.B -slotinfo
displays long form
.I pups(3)
library dependencies.
.br

.B -argf
.I argument_file
tells the filter to take its command line arguments from the specified
.I argument_file.
The data format of the argument file is:
.br

#  ... optional comment line ...
.br
.B -argflag           [1]
.br
.I argument_parameter [1,1]
.br
.I argument_parameter [1,2]
.br

.I argument_parameter [1,n]
.br

# ... optional comment line ...
.br
.B -argflag           [2]
.br
.I argument_parameter [1,1]
.br
.I argument_parameter [1,2]
.br

.I argument_parameter [1,n]
.br
 
.B -nice
.I cardinal
determimes the
.I niceness
at which the filter is scheduled. In keeping with all UNIX user processes, nice
must be a cardinal between 0 and 20. The default niceness for
.I xcat(1)
is 4.
.br

.B -usage
displays program shortform usage.
.br

.B -version
displays program version number.
.br

.SH PSRP OPTIONS

.B dataflow
.I off | stdout | psrp | ptap
switches the output of
.I xcat(1)
to the specified stream. If
.B ptap
is specified output is sent to both
.B standard output
and
.B psrp
streams.
.b r

.SH LIBRARIES
.I xcat(1)
uses the PUPS libraries, see
.I pups(3)
.br

.SH EXIT STATUS

255: error
.br
1:
.B -usage
or
.B -help
.br
0: no error
.br

.SH ENVIRONMENTAL VARIABLES
none.
.br

.SH COPYRIGHT
(C) Tumbling Dice Ltd, Gosforth.
.br

.SH PROGRAMMER
M.A. O'Neill.
.br


pupscore.src/xcat.c
/*------------------------------------------------------------------------------
   Purpose: Place marker for monitoring applications in live pipelines 

    Author:  M.A. O'Neill
             Tumbling Dice Ltd
             Gosforth
             Newcastle upon Tyne
             NE3 4RT
             United Kingdom

    Version: 2.02 
    Dated:   4th January 2022
    E-mail:  mao@tumblingdice.co.uk 
------------------------------------------------------------------------------*/

#include <me.h>
#include <utils.h>
#include <signal.h>
#include <string.h>
#include <psrp.h>
#include <unistd.h>
#include <vstamp.h>


/*-----------------*/
/* Version of xcat */
/*-----------------*/

#define XCAT_VERSION    "2.02"


#ifdef BUBBLE_MEMORY_SUPPORT
#include <bubble.h>
#endif /* BUBBLE_MEMORY_SUPPORT */


/*----------------------------------------------------------------*/
/* Get signal mapping appropriate to OS and hardware architecture */
/*----------------------------------------------------------------*/

#define __DEFINE__
#if defined(I386) || defined(X86_64)
#include <sig.linux.x86.h>
#endif /* I386 || X86_64 */

#ifdef ARMV6L
#include <sig.linux.arm.h>
#endif /* ARMV6L */

#ifdef ARMV7L
#include <sig.linux.arm.h>
#endif /* ARMV7L */

#ifdef AARCH64
#include <sig.linux.arm.h>
#endif /* AARCH64 */
#undef __DEFINE__




/*-----------------------------------------------------------------------------*/
/* Version ID (used by CKPT enabled binaries to discard stale checkpoint files */
/*-----------------------------------------------------------------------------*/

#include <vstamp.h>


/*------------------------------------------------------------------------------
    Get application information for slot manager ...
------------------------------------------------------------------------------*/
/*---------------------------*/
/* Slot information function */
/*---------------------------*/

_PRIVATE void xcat_slot(int level)
{   (void)fprintf(stderr,"int app xcat %s: [ANSI C]\n",XCAT_VERSION);

    if(level > 1)
    {  (void)fprintf(stderr,"(C) 2005-2022 Tumbling Dice\n");
       (void)fprintf(stderr,"Author: M.A. O'Neill\n");
       (void)fprintf(stderr,"Pipeline insertion marker for dynamic pipeline monitors (built %s %s)\n\n",__TIME__,__DATE__);
    }
    else
       (void)fprintf(stderr,"\n");
    (void)fflush(stderr);  
}
					 



/*----------------------------*/
/* Application usage function */
/*----------------------------*/

_PRIVATE void xcat_usage(void)

{   (void)fprintf(stderr,"xcat [< <input file> > <output file> >& <ASCII log file>]\n");

    (void)fprintf(stderr,"\n\nSignals\n\n");
    (void)fprintf(stderr,"SIGINIT SIGCHAN SIGPSRP: Process status [PSRP] request (protocol %4.2F)\n",PSRP_PROTOCOL_VERSION);
    (void)fprintf(stderr,"SIGCLIENT: tell client server is about to segment\n");

    #ifdef CRUI_SUPPORT
    (void)fprintf(stderr,"SIGCHECK SIGRESTART:      checkpoint and restart signals\n");
    #endif /* CRIU_SUPPORT */

    (void)fprintf(stderr,"SIGALIVE: check for existence of client on signal dispatch host\n");
    (void)fflush(stderr);
}
		    

#ifdef SLOT
#include <slotman.h>
_EXTERN void (* SLOT)() __attribute__ ((aligned(16))) = xcat_slot;
_EXTERN void (* USE )() __attribute__ ((aligned(16))) = xcat_usage;
#endif /* SLOT */


/*------------------------------------------------------------------------------
    Application build date ...
------------------------------------------------------------------------------*/

_EXTERN char appl_build_time[SSIZE] = __TIME__;
_EXTERN char appl_build_date[SSIZE] = __DATE__;





/*------------------------------------------------------------------------------
    Defines which are local to this application ...
------------------------------------------------------------------------------*/

#define TO_STDOUT    1
#define TO_PSRP      2
#define TO_PTAP      4
#define TO_DATASINK  8




/*------------------------------------------------------------------------------
    Private variables required by this application ...
------------------------------------------------------------------------------*/

                                                       /*-------------------------------------*/
_PRIVATE _BOOLEAN data_flowing         = TRUE;         /* State of dataflow (stdin -> stdout) */
_PRIVATE _BOOLEAN root_instance        = TRUE;         /* State of dataflow (stdin -> stdout) */
_PRIVATE int      bytes_read;                          /* Number of bytes read from stdin     */
_PRIVATE char     line_of_input[SSIZE] = "";           /* Line buffer                         */
_PRIVATE char     mytmpfile[SSIZE]     = "";           /* Temporary file name                 */
_PRIVATE FILE     *mystream            = (FILE *)NULL; /* I/O stream                          */
                                                       /*-------------------------------------*/



/*------------------------------------------------------------------------------
    Local prototype function declarations ...
------------------------------------------------------------------------------*/

/* Report status of function to PSRP client */
_PROTOTYPE _PRIVATE int psrp_process_status(int, char *[]);

/* Switch dataflow on/off */
_PROTOTYPE _PRIVATE int dataflow(int, char *[]);




/*------------------------------------------------------------------------------
    Report process status ...
------------------------------------------------------------------------------*/

_PRIVATE int psrp_process_status(int argc, char *argv[])

{    (void)fprintf(psrp_out,"\n    Pipleline dynamic insertion marker status\n");
     (void)fprintf(psrp_out,"    =========================================\n\n");
     (void)fflush(psrp_out); 

     #if defined(CRIU_SUPPORT)
     (void)fprintf(psrp_out,"    Binary is Criu enabled (checkpointable)\n");
     #endif /* CRIU_SUPPORT */

     if(data_flowing == TO_STDOUT)
        (void)fprintf(psrp_out,"    xcat (%d@%s): transferring data (from stdin to stdout)\n",appl_pid,appl_host);
     else if(data_flowing == TO_DATASINK)
        (void)fprintf(psrp_out,"    xcat (%d@%s): dataflow interrupted\n",appl_pid,appl_host);

     (void)fprintf(psrp_out,"\n\n");
     (void)fflush(psrp_out);

     return(PSRP_OK);
}




/*------------------------------------------------------------------------------
    Switch dataflow to specified sink ...
------------------------------------------------------------------------------*/

_PRIVATE int dataflow(int argc, char *argv[])

{   if(strcmp(argv[1],"help") == 0 || strcmp(argv[1],"usage") == 0)
    {  (void)fprintf(psrp_out,"usage: dataflow [stdout | psrp | ptap | off]\n");
       (void)fflush(psrp_out);

       return(PSRP_OK);
    }

    if(argc == 1)
    {  if(data_flowing == TO_DATASINK)
          (void)fprintf(psrp_out,"\n    dataflow: OFF\n");
       else if(data_flowing == TO_STDOUT)
          (void)fprintf(psrp_out,"\n    dataflow: standard output\n");
    }
    else if(strcmp(argv[1],"stdout")  == 0)
    {  data_flowing = TO_STDOUT;

       if(appl_verbose == TRUE)
       {  (void)strdate(date);
          (void)fprintf(stderr,"%s %s (%d@%s:%s): output redirected to standard output\n",
                                             date,appl_name,appl_pid,appl_host,appl_owner); 
          (void)fflush(stderr);
       }
    }
    else if(strcmp(argv[1],"psrp")  == 0 || strcmp(argv[1],"ptap") == 0)
    {  char xcat_command[SSIZE] = "";

       if(strcmp(argv[1],"psrp")  == 0)
          data_flowing = TO_PSRP;
       else
          data_flowing = TO_PTAP;

       if(appl_verbose == TRUE)
       {  (void)strdate(date);

          if(strcmp(argv[1],"psrp")  == 0)
             (void)fprintf(stderr,"%s %s (%d@%s:%s): output redirected to PSRP client\n",
                                            date,appl_name,appl_pid,appl_host,appl_owner); 
          else
             (void)fprintf(stderr,"%s %s (%d@%s:%s): output tapped to PSRP client\n",
                                        date,appl_name,appl_pid,appl_host,appl_owner); 
          (void)fflush(stderr);
       }


       /*-------------------------------------------------*/
       /* Client side abort callback for xcat application */
       /*-------------------------------------------------*/

       do {   bytes_read = read(0,line_of_input,SSIZE);

              if(bytes_read > 0)
              {  (void)fputs(line_of_input,psrp_out);
                 (void)fflush(psrp_out);


                 /*------------------------------------------------*/
                 /* If we are tapping standard output send data to */
                 /* standard output as well.                       */
                 /*------------------------------------------------*/

                 if(strcmp(argv[1],"ptap")  == 0 && root_instance == TRUE)
                    (void)write(1,line_of_input,bytes_read);
              }
          } while(bytes_read > 0);

       (void)fprintf(psrp_out,"\nEnd of file (on standard input)\n\n");
       (void)fflush(psrp_out);

       pups_exit(0);
    }
    else if(strcmp(argv[1],"off")  == 0)
    {  data_flowing = TO_DATASINK;

       if(appl_verbose == TRUE)
       {  (void)strdate(date);
          (void)fprintf(stderr,"%s %s (%d@%s:%s): output redirected to data sink\n",
                                       date,appl_name,appl_pid,appl_host,appl_owner); 
          (void)fflush(stderr);
       }
    }

    return(PSRP_OK);
}



/*------------------------------------------------------------------------------
    Remove junk ...
------------------------------------------------------------------------------*/

_PRIVATE int remove_junk(void)

{    (void)pups_fclose(mystream);
     (void)unlink(mytmpfile);
}




/*------------------------------------------------------------------------------
    Software I.D. tag (used if CKPT support enabled to discard stale dynamic
    checkpoint files) ...
------------------------------------------------------------------------------*/

#define VTAG  3700

extern int appl_vtag = VTAG;





/*------------------------------------------------------------------------------
    Main entry point ...
------------------------------------------------------------------------------*/

_PUBLIC int pups_main(int argc, char *argv[])


{

    /*--------------------------------*/
    /* Wait for PUPS/P3 to initialise */
    /*--------------------------------*/

    (void)psrp_ignore_requests();


    /*--------------------*/
    /* Parse command tail */
    /*--------------------*/


    /*-----------------------------*/
    /* Do standard initialisations */
    /*-----------------------------*/

    pups_std_init(TRUE,
                  &argc,
                  XCAT_VERSION,
                  "M.A. O'Neill",
                  "xcat",
                  "2022",
                  argv);

    (void)psrp_init(PSRP_STATUS_ONLY  | PSRP_HOMEOSTATIC_STREAMS,&psrp_process_status);
    (void)psrp_attach_static_function("status",(void *)&psrp_process_status);
    (void)psrp_attach_static_function("flow",(void *)&dataflow);
    (void)psrp_load_default_dispatch_table();
    (void)psrp_accept_requests();

    if(root_instance == TRUE)
       (void)pups_register_exit_f("remove_junk",&remove_junk,(char *)NULL);


    /*------------------------------------------*/ 
    /* We must disable automatic child handling */
    /*------------------------------------------*/ 

    pups_noauto_child();


    /*-------------------------------------------*/
    /* Check command tail for unparsed arguments */
    /*-------------------------------------------*/

    pups_t_arg_errs(argd,args);

    if(isatty(0) == 1)
       pups_error("[xcat] no file attached to stdin");


    /*--------------------------------------------------------------------------*/
    /* This is the pups_main part of the program - it takes data from stdin and */
    /* pushes it to stdout                                                      */
    /*--------------------------------------------------------------------------*/

    do {   
           if(data_flowing == TO_PSRP || data_flowing == TO_PTAP)
           {  data_flowing = TO_STDOUT;

              if(appl_verbose == TRUE)
              {  (void)strdate(date);
                 (void)fprintf(stderr,"%s %s (%d@%s:%s): dataflow to standard output resumed\n",
                                                   date,appl_name,appl_pid,appl_host,appl_owner);
                 (void)fflush(stderr);
              }

              if(bytes_read > 0 && root_instance == TRUE)
                 (void)pups_write(1,line_of_input,bytes_read);
           }

           bytes_read = read(0,line_of_input,SSIZE);
           if(bytes_read > 0)
           {  if(data_flowing == TO_STDOUT)
                 (void)pups_write(1,line_of_input,bytes_read);
           }
       } while(bytes_read > 0);

    pups_exit(0);
}
